linalg/slepc.cpp

```diff
@@ -0,0 +1,342 @@
+   MPI_Comm comm;
+   ierr = PetscObjectGetComm((PetscObject)eps,&comm); PCHKERRQ(eps,ierr);
+   ierr = EPSDestroy(&eps); CCHKERRQ(comm,ierr);
+}
+
+
+void SlepcEigenSolver::SetOperator(const Operator &op)
+{
+   PetscParMatrix *pA = const_cast<PetscParMatrix *>
+                        (dynamic_cast<const PetscParMatrix *>(&op));
+   const HypreParMatrix *hA = dynamic_cast<const HypreParMatrix *>(&op);
+   const Operator       *oA = dynamic_cast<const Operator *>(&op);
+   bool delete_pA = false;
+
+   if (!pA)
+   {
+      if (hA)
+      {
+         pA = new PetscParMatrix(hA,
+                                 _wrap ? Operator::PETSC_MATSHELL : Operator::PETSC_MATAIJ);
+         delete_pA = true;
+      }
+      else if (oA)
+      {
+         pA = new PetscParMatrix(PetscObjectComm((PetscObject)eps),oA,
+                                 _wrap ? Operator::PETSC_MATSHELL : Operator::PETSC_MATAIJ);
+         delete_pA = true;
+
+      }
+   }
+   MFEM_VERIFY(pA, "Unsupported operation!");
+
+   if (operatorset)
+   {
+      delete VR;
+      delete VC;
+      VR = VC = NULL;
+   }
+   ierr = EPSSetOperators(eps,*pA,NULL); PCHKERRQ(eps, ierr);
+   operatorset = true;
+
+   if (delete_pA) {delete_pA;}
+}
+
+void SlepcEigenSolver::SetOperators(const Operator &op, const Operator &opB)
+{
+   PetscParMatrix *pA = const_cast<PetscParMatrix *>
+                        (dynamic_cast<const PetscParMatrix *>(&op));
+   PetscParMatrix *pB = const_cast<PetscParMatrix *>
+                        (dynamic_cast<const PetscParMatrix *>(&opB));
+   const HypreParMatrix *hA = dynamic_cast<const HypreParMatrix *>(&op);
+   const HypreParMatrix *hB = dynamic_cast<const HypreParMatrix *>(&opB);
+
+   const Operator *oA = dynamic_cast<const Operator *>(&op);
+   const Operator *oB = dynamic_cast<const Operator *>(&opB);
+   bool delete_pA = false;
+   bool delete_pB = false;
+   if (!pA)
+   {
+      if (hA)
+      {
+         pA = new PetscParMatrix(hA,
+                                 _wrap ? Operator::PETSC_MATSHELL : Operator::PETSC_MATAIJ);
+         delete_pA = true;
+      }
+      else if (oA)
+      {
+         pA = new PetscParMatrix(PetscObjectComm((PetscObject)eps),oA,
+                                 _wrap ? Operator::PETSC_MATSHELL : Operator::PETSC_MATAIJ);
+         delete_pA = true;
+      }
+   }
+   MFEM_VERIFY(pA, "Unsupported Operation!");
+   if (!pB)
+   {
+      if (hB)
+      {
+         pB = new PetscParMatrix(hB,
+                                 _wrap ? Operator::PETSC_MATSHELL : Operator::PETSC_MATAIJ);
+         delete_pB = true;
+      }
+      else if (oB)
+      {
+         pB = new PetscParMatrix(PetscObjectComm((PetscObject)eps),oB,
+                                 _wrap ? Operator::PETSC_MATSHELL : Operator::PETSC_MATAIJ);
+         delete_pB = true;
+      }
+   }
+   MFEM_VERIFY(pB, "Unsupported Operation!");
+
+   if (operatorset)
+   {
+      delete VR;
+      delete VC;
+      VR = VC = NULL;
+   }
+   operatorset = true;
```

Comment:
`operatorset` is used in the `PetscLinearSolver` class because we can in principle reuse the `KSP` with different matrix sizes, but we need to call `KSPReset` on it. I'm not sure we can do the same with the `EPS` object. I suggest you remove this logic.

Reply:
Another difference between `EPS` and `KSP`, is that `EPSSetOperators` calls `EPSReset` if necessary. However we still need to reset VC and VR is we reuse the `EPS`. I could remove the logic and still always delete `VC` and `VR`?

Reply:
Yes, delete and recreate the vectors placeholders at each call to SetOperators then.

Reply:
Done